py/nlr.h

```diff
@@ -30,71 +30,69 @@
 // exception handling, basically a stack of setjmp/longjmp buffers
 
 #include <limits.h>
-#include <setjmp.h>
 #include <assert.h>
 
 #include "py/mpconfig.h"
 
-typedef struct _nlr_buf_t nlr_buf_t;
-struct _nlr_buf_t {
-    // the entries here must all be machine word size
-    nlr_buf_t *prev;
-    void *ret_val; // always a concrete object (an exception instance)
-#if !defined(MICROPY_NLR_SETJMP) || !MICROPY_NLR_SETJMP
+// If MICROPY_NLR_SETJMP is not enabled then auto-detect the machine arch
+// Allow a port to set MICROPY_NLR_NUM_REGS to define their own implementation
+#if !MICROPY_NLR_SETJMP && !defined(MICROPY_NLR_NUM_REGS)
 #if defined(__i386__)
-    void *regs[6];
+    #define MICROPY_NLR_X86 (1)
+    #define MICROPY_NLR_NUM_REGS (6)
 #elif defined(__x86_64__)
-  #if defined(__CYGWIN__)
-    void *regs[12];
-  #else
-    void *regs[8];
-  #endif
```

Comment:
@stinos furthermore: for `__CYGWIN__` why is it 12 regs and not 10 (an extra 2 for rsi and rdi compared with non-unix)?  And why isn't this check `defined(_WIN32) || defined(__CYGWIN__)` to match the definition of the `NLR_OS_WINDOWS` macro?

Reply:
Not sure. This was introduced years ago by you and it didn't change since. However I think this might actually have been an oversight in #1632 and follow up bf1570cb2559ba12ca86b2dbb26be665300ffe5e: there the _WIN32 got introduced but nlr.h has not been changed. And this seems very wrong (not tested, but in nlrx64 I see more movq calls than there is space in nlr->regs, right?) So that might even be the reason why 64bit msys mingw builds segfault (assuming they still do). Not sure what the best way to solve this would be now: I don't have time to test all of this right now, though I do next week. So either you leave it as-is and we'll see later one, or you do something like declaring NLR_OS_WINDOWS in nlr.h so it can be used both there to enable set MICROPY_NLR_NUM_REGS to 12 and in nlr.c?

Reply:
> This was introduced years ago by you and it didn't change since

Indeed!  And it looks like it allocates 2 more slots than it needs to.

> there the _WIN32 got introduced but nlr.h has not been changed. And this seems very wrong (not tested, but in nlrx64 I see more movq calls than there is space in nlr->regs, right?)

Yes it seems wrong.

Ok, let's leave this particular issue until later.  For now I've aimed to make this PR not change any behaviour, as far as that is possible.